File: go/test/heat_stress_warning_en_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/heat-stress-warning-sdk"
	"github.com/voxgig-sdk/heat-stress-warning-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestHeatStressWarningEnEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.HeatStressWarningEn(nil)
		if ent == nil {
			t.Fatal("expected non-nil HeatStressWarningEnEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := heat_stress_warning_enBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "heat_stress_warning_en." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		heatStressWarningEnRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.heat_stress_warning_en", setup.data)))
		var heatStressWarningEnRef01Data map[string]any
		if len(heatStressWarningEnRef01DataRaw) > 0 {
			heatStressWarningEnRef01Data = core.ToMapAny(heatStressWarningEnRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = heatStressWarningEnRef01Data

		// LIST
		heatStressWarningEnRef01Ent := client.HeatStressWarningEn(nil)
		heatStressWarningEnRef01Match := map[string]any{}

		heatStressWarningEnRef01ListResult, err := heatStressWarningEnRef01Ent.List(heatStressWarningEnRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, heatStressWarningEnRef01ListOk := heatStressWarningEnRef01ListResult.([]any)
		if !heatStressWarningEnRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", heatStressWarningEnRef01ListResult)
		}

	})
}

func heat_stress_warning_enBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "heat_stress_warning_en", "HeatStressWarningEnTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read heat_stress_warning_en test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse heat_stress_warning_en test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"heat_stress_warning_en01", "heat_stress_warning_en02", "heat_stress_warning_en03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID": idmap,
		"HEATSTRESSWARNING_TEST_LIVE":      "FALSE",
		"HEATSTRESSWARNING_TEST_EXPLAIN":   "FALSE",
		"HEATSTRESSWARNING_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["HEATSTRESSWARNING_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["HEATSTRESSWARNING_APIKEY"],
			},
			extra,
		})
		client = sdk.NewHeatStressWarningSDK(core.ToMapAny(mergedOpts))
	}

	live := env["HEATSTRESSWARNING_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["HEATSTRESSWARNING_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/heat_stress_warning_en_entity_test.rb
# HeatStressWarningEn entity test

require "minitest/autorun"
require "json"
require_relative "../HeatStressWarning_sdk"
require_relative "runner"

class HeatStressWarningEnEntityTest < Minitest::Test
  def test_create_instance
    testsdk = HeatStressWarningSDK.test(nil, nil)
    ent = testsdk.HeatStressWarningEn(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = heat_stress_warning_en_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "heat_stress_warning_en." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    heat_stress_warning_en_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.heat_stress_warning_en")))
    heat_stress_warning_en_ref01_data = nil
    if heat_stress_warning_en_ref01_data_raw.length > 0
      heat_stress_warning_en_ref01_data = Helpers.to_map(heat_stress_warning_en_ref01_data_raw[0][1])
    end

    # LIST
    heat_stress_warning_en_ref01_ent = client.HeatStressWarningEn(nil)
    heat_stress_warning_en_ref01_match = {}

    heat_stress_warning_en_ref01_list_result = heat_stress_warning_en_ref01_ent.list(heat_stress_warning_en_ref01_match, nil)
    assert heat_stress_warning_en_ref01_list_result.is_a?(Array)

  end
end

def heat_stress_warning_en_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "heat_stress_warning_en", "HeatStressWarningEnTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = HeatStressWarningSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["heat_stress_warning_en01", "heat_stress_warning_en02", "heat_stress_warning_en03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID" => idmap,
    "HEATSTRESSWARNING_TEST_LIVE" => "FALSE",
    "HEATSTRESSWARNING_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["HEATSTRESSWARNING_TEST_HEAT_STRESS_WARNING_EN_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["HEATSTRESSWARNING_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = HeatStressWarningSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["HEATSTRESSWARNING_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["HEATSTRESSWARNING_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
